packages/pubsub/src/connection-pool.js

```diff
@@ -0,0 +1,416 @@
+    });
+
+    self.connections.set(id, connection);
+  });
+};
+
+/**
+ * Creates a message object for the user.
+ *
+ * @param {string} connectionId - The connection id that the message was
+ *     received on.
+ * @param {object} resp - The message response data from StreamingPull.
+ * @return {object} message - The message object.
+ */
+ConnectionPool.prototype.createMessage = function(connectionId, resp) {
+  var self = this;
+
+  var pt = resp.message.publishTime;
+  var milliseconds = parseInt(pt.nanos, 10) / 1e6;
+  var originalDataLength = resp.message.data.length;
+
+  return {
+    connectionId: connectionId,
+    ackId: resp.ackId,
+    id: resp.message.messageId,
+    attributes: resp.message.attributes,
+    publishTime: new Date(parseInt(pt.seconds, 10) * 1000 + milliseconds),
+    received: Date.now(),
+    data: resp.message.data,
+    // using get here to prevent user from overwriting data
+    get length() {
+      return originalDataLength;
+    },
+    ack: function() {
+      self.subscription.ack_(this);
+    },
+    nack: function() {
+      self.subscription.nack_(this);
+    }
+  };
+};
+
+/**
+ * Gets the Subscriber client. We need to bypass GAX until they allow deadlines
+ * to be optional.
+ *
+ * @param {function} callback - The callback function.
+ * @param {?error} callback.err - An error occurred while getting the client.
+ * @param {object} callback.client - The Subscriber client.
+ */
+ConnectionPool.prototype.getClient = function(callback) {
+  if (this.client) {
+    callback(null, this.client);
+    return;
+  }
+
+  var self = this;
+  var pubsub = this.subscription.pubsub;
+
+  pubsub.auth.getAuthClient(function(err, authClient) {
+    if (err) {
+      callback(err);
+      return;
+    }
+
+    var credentials = grpc.credentials.combineChannelCredentials(
+      grpc.credentials.createSsl(),
+      grpc.credentials.createFromGoogleCredential(authClient)
+    );
+
+    var Subscriber = v1(pubsub.options).Subscriber;
+
+    self.client = new Subscriber(v1.SERVICE_ADDRESS, credentials, {
+      'grpc.max_receive_message_length': 20000001,
+      'grpc.primary_user_agent': common.util.getUserAgentFromPackageJson(PKG)
+    });
+
+    callback(null, self.client);
+  });
+};
+
+/**
+ * Check to see if at least one stream in the pool is connected.
+ * @return {boolean}
+ */
+ConnectionPool.prototype.isConnected = function() {
+  var interator = this.connections.values();
+  var connection = interator.next().value;
+
+  while (connection) {
+    if (connection.isConnected) {
+      return true;
+    }
+
+    connection = interator.next().value;
+  }
+
+  return false;
+};
+
+/**
+ * Creates specified number of connections and puts pool in open state.
+ */
+ConnectionPool.prototype.open = function() {
+  var existing = this.connections.size;
+  var max = this.settings.maxConnections;
+
+  for (; existing < max; existing++) {
+    this.queueConnection();
+  }
+
+  this.isOpen = true;
```

Comment:
Shouldn't we set self.noConnectionsTime to Date.now() here? Don't we want to "start our timer" as soon as open() is called? Because we have 0 connections at this very moment, and the user has declared an intent to receive messages. If we wait for some future event before we start our timer, what is there a bound on that wait? If there is no bound, then we might have 0 connections forever and never report it to the user.